sources/xfscli/src/bin/xfs_admin.rs

```diff
@@ -0,0 +1,212 @@
+        io_opts.append(&mut vec!["-r", "-c", "fsuuid"]);
+    }
+
+    let mut uuid_arg = "uuid ".to_string();
+    if let Some(uuid) = &args.uuid {
+        uuid_arg.push_str(uuid.as_str());
+        db_opts.append(&mut vec!["-c", uuid_arg.as_str()]);
+    }
+
+    if let Some(device) = &args.device {
+        // Exit if filesystem is mounted and require offline for queried arguments
+        if let Some(mpt) = get_mount_point(device) {
+            // filesystem is mounted
+            if is_required_offline(&args) {
+                println!("filesystem is mounted {}", mpt);
+                exit(2);
+            }
+
+            if !io_opts.is_empty() {
+                let mut xfs_io = Command::new(XFS_IO);
+                xfs_io.args(["-p", "xfs_admin"]);
+                xfs_io.args(io_opts);
+                xfs_io.arg(mpt.as_str().trim());
+
+                let xfs_io_result = xfs_io
+                    .output()
+                    .expect("Couldn't run xfs_io for given device");
+                print!("{}", String::from_utf8(xfs_io_result.stdout).unwrap());
+                exit(xfs_io_result.status.code().unwrap());
+            }
+        }
+
+        let mut status = 0;
+        if !db_opts.is_empty() {
+            let mut xfs_db_args = vec!["-x", "-p", "xfs_admin"];
+
+            let mut log_location_arg = "-l ".to_string();
+            if let Some(log_location) = &args.log_location {
+                log_location_arg.push_str(log_location.as_str());
+                xfs_db_args.push(log_location_arg.as_str());
+            }
+
+            xfs_db_args.append(&mut db_opts);
+            xfs_db_args.push(device);
+
+            status = query_xfs_db(xfs_db_args);
+        }
+
+        if !repair_opts.is_empty() {
+            print!("Running xfs_repair to upgrade filesystem.");
+            let mut repair_args: Vec<&str> = vec![];
+
+            let mut log_location_arg = "-l ".to_string();
+            if let Some(log_location) = &args.log_location {
+                log_location_arg.push_str(log_location.as_str());
+                repair_args.push(log_location_arg.as_str());
+            }
+
+            let mut realtime_device_arg = "-r ".to_string();
+            if let Some(realtime_device) = &args.realtime_device {
+                realtime_device_arg.push_str(realtime_device);
+                repair_args.push(realtime_device_arg.as_str());
+            }
+
+            repair_args.append(&mut repair_opts);
+            repair_args.push(device.as_str());
+
+            status += perform_xfs_repair(repair_args);
+        }
+        exit(status);
+    } else {
+        println!("Couldn't parse target block-device");
+        exit(2);
+    }
+}
+
+// Filesystem should not be mounted with args in this function
+// For more reference check require_offline variable in
+// https://git.kernel.org/pub/scm/fs/xfs/xfsprogs-dev.git/tree/db/xfs_admin.sh
+fn is_required_offline(args: &Args) -> bool {
```

Comment:
Why not making this a function of `Args`? so that you can call `args.is_required_offline`?